src/main/java/sbt_inc/SbtIncrementalCompiler.java

```diff
@@ -89,18 +91,35 @@ public SbtIncrementalCompiler(
 
     List<File> allJars = new ArrayList<>(extraJars);
     allJars.add(libraryJar);
-    allJars.add(reflectJar);
+    if (reflectJar != null) {
+      allJars.add(reflectJar);
+    }
+    if (interfacesJar != null) {
+      allJars.add(interfacesJar);
+    }
     allJars.add(compilerJar);
 
+    URL[] urls =
+        reflectJar != null
+            ? new URL[] {
+              libraryJar.toURI().toURL(), reflectJar.toURI().toURL(), compilerJar.toURI().toURL()
+            }
+            : interfacesJar != null
+                ? new URL[] {
+                  libraryJar.toURI().toURL(),
+                  interfacesJar.toURI().toURL(),
+                  compilerJar.toURI().toURL()
+                }
+                : new URL[] {libraryJar.toURI().toURL(), compilerJar.toURI().toURL()};
+
+    URLClassLoader classLoader =
+        scalaVersion.major == 3
+            ? new scala_maven.Scala3CompilerLoader(urls, xsbti.Reporter.class.getClassLoader())
+            : new URLClassLoader(urls);
     ScalaInstance scalaInstance =
         new ScalaInstance(
             scalaVersion.toString(), // version
-            new URLClassLoader(
-                new URL[] {
-                  libraryJar.toURI().toURL(),
-                  reflectJar.toURI().toURL(),
-                  compilerJar.toURI().toURL()
-                }), // loader
+            classLoader, // loader
```

Comment:
The error indicates that `scala-library` is missing from the classpath.

The `ScalaInstance` constructor that is used here is deprecated. You should use the one defined here: https://github.com/sbt/zinc/blob/a096e0d32dd46c0be4c831a37b497947e7479b3a/internal/zinc-classpath/src/main/scala/sbt/internal/inc/ScalaInstance.scala#L34-L43

Make sure that the different lists of jars are correct:
`compilerJars` should contain:
- `scala3-compiler`
- `scala3-interfaces`
- ` tasty-core`
- ...

`libraryJars` should contain:
- `scala3-library`
-  `scala-library` (the scala3-library depends on the Scala 2.13 library)

`allJars` should contain all of them.

Note that each list of jars is associated with a classloader:
```
val loaderLibraryOnly = new Scala3CompilerLoader(libraryJars, xsbti.Reporter.class.getClassLoader())
val loaderCompilerOnly = new URLClassLoader(compilerJars, loaderLibraryOnly)
val loader = loaderCompilerOnly
```

Reply:
As you can see the logic to configure the Scala 3 compiler is quite different from Scala 2. Maybe you should have 2 different constructors.